### PR TITLE
Simplifies the boolean-bitfield implementation to use `bit-vec` crate

### PR DESCRIPTION
**Pull request by ralexstokes**
## Issue Addressed

#22 

## Proposed Changes

Simplifies the implementation of `boolean-bitfield` crate by using the `bit-vec` crate which handles most of the functionality for us.

## Additional Info

Opening this to start the integration process; marked WIP until this changes.

Some notes/questions:

Usage: will we ever be starting with an empty bitfield and expecting to use the API to insert additional bits? There are handy `push/pop` fns for `bit-vec` but they aren't exposed right now. We should also note that for now, `BooleanBitfield::new()` returns an empty bitset and we can't do anything with it. Bringing this up because I expect this is not what we ultimately want.

API question: I implemented the API given on #22 although I have included a `len` method which returns the number of bits this bitfield is tracking. The API on the issue requests the number of backing bytes. Given that `bit-vec` tracks the bit count I figure we can just use this. If we really want a byte count then I can do the `|x| (x + 7) // 8` trick.

ssz and canonical encoding: if you can point me to some test vectors for ssz {en,de}coding of bitfields then I'll add tests here. in particular, it looks like `bit-vec` uses big endian encoding so we will need to make sure that matches what is coming off the wire and/or adjust accordingly.

### CODE REVIEW

**Comment by ralexstokes** [comment]
Oh also, what is the repo policy on `rustfmt`. This is the one style thing I'll be vocal about and otherwise defer to the repo policy. I've withheld a messy diff that was just a bunch of `rustfmt` changes on a file i touched.

I would expect we will want to go through and `rustfmt` every file in one big commit if we need to do that. i'm happy to do that if there is interest

**Comment by paulhauner** [comment]
Hey, thanks for this! I'm still dealing with some backlog, but I'll jump onto this in the coming days.

I agree about using `rustfmt`. Thanks for bringing it up, my vim plugin must have died at some point. I'll check it out and make an issue to do a big `rustfmt` sweep :)

**Comment by paulhauner** [comment]
Finally got around to checking this out! I like the bitvec crate, seems like a good idea.

I pulled this branch and ran the cargo tests, some failed. I guess you're aware of that :)

Let me know how you want to proceed with this -- do you need to do some more work to take it out of WIP stage?

**Comment by ralexstokes** [comment]
@paulhauner it actually looks like some of this was answered as i went in to fix the tests :)

i'll push what i have but need to figure out what the "zero value" for a bitfield is before the `ssz` tests are passing....

is the empty bitfield "nothing" or the byte `0`. in the previous impl, you have `vec![0]` as the default whereas the current impl has `vec![]`. this is causing the block hash to differ and the test fails...

as soon as you verify which is the correct option then i'll update!

**Comment by paulhauner** [comment]
Hey @ralexstokes! Once again, apologies for the delayed responses.

The zero value should be a zero byte `0x00`. I can't remember where, but I remember @djrtwo stating this :)

**Comment by ralexstokes** [comment]
I would be curious to hear the rationale because it would seem much easier if the length of the field corresponded to how many validators are in the set. But I suppose there isn't any part of the bitfield's lifetime where this is an important distinction to make.

**Comment by ralexstokes** [comment]
@paulhauner seems like the serialization is passing now... but another test is failing and leads to another question of semantics :)

in particular, `signature_verification::tests::test_signature_verification` is failing because of the behavior of `verify_aggregate_signature_for_indices`.

We start with a default bitfield with length `8` and set the first `6` bits to `true`. Then in the function we scan across `12` bits to see who voted (following the `len()` of the `attestation_indices`). When we get to the first bit that doesn't exist in the field (the 9th bit), we return an `Err(OutOfBounds)` which the test panics on.

Following the test code here as an example, it seems that we want the length of the field to be `12` (or rounded up to the nearest byte) but there is no way to indicate that to the field as it stands. If we always know the max size we will have ahead of time we can pass that in as a parameter. Otherwise we will have to implement some re-sizing logic and at that point the notion of "out of bounds" becomes murky.

Do we always know the maximum size at any point we create a bitfield? What happens if we create a default bitfield (which has length 8) and then need to extend at a later point in time (this scenario may not ever occur if the previous question is a yes)?

**Comment by rawfalafel** [comment]
@ralexstokes , @paulhauner  I hate to chime in like this on an open PR :smiling_imp: but have you considered decoding the bitfield into a boolean array instead? Of course its larger, but the implementation becomes way simpler and probably faster. The spec uses a bitfield for space efficiency, but there's no reason for the implementation to handle it as the same data structure.

**Comment by ralexstokes** [comment]
@rawfalafel that is definitely one option. the spec uses a bitfield for space complexity and it is in line w/ the goals of lighthouse (@paulhauner can chime in if otherwise) to also strive for this type of efficiency :)

i'm personally not that concerned as we have this very handy `bit-vec` crate that does all of the bit-twiddling for us. So we minimize how much memory we are using. And in terms of time complexity aka speed -- as far as I have seen the underlying implementation of `bit-vec` basically uses the same infrastructure for that we would use for a boolean array anyway -- the `Vec<T>` type. If we were at the stage where we needed to squeeze out even more perf and had benchmarked that this particular use of the data structure was a hotspot then we could do some profiling to see what is better. My current understanding is that it would be about the same perf-wise and we don't have to write any packing/unpacking logic to {de,}serialize to the wire.

**Comment by paulhauner** [comment]
Hey @ralexstokes -- good point about the tests expecting an extensible bitfield. I think it's fine to have an extensible bitfield. If we know the length during initialization we can just use `with_capacity` to avoid re-allocations.

In production code I can't think of a scenario where we _don't_ know the length beforehand, but this might be the case when developing tests. Most importantly, I can't think of a scenario where it's dangerous to extend a bitfield or expect it to be longer than it actually is. My call would be to go ahead and make it extensible, sorry for not thinking of this earlier!

@rawfalafel thanks for the input, always happy to hear from you!

In Rust, array sizes must be known at compile time. This restriction means that we can't decode into an array, we'd need to decode into a vec (unless we want to allocate a maximally-sized array each time). The crate we're using ([bit-vec](https://github.com/contain-rs/bit-vec)) a vec of `uint32` for, as you mentioned, space efficiency.

The things we're discussing presently are the semantics about whether or not a bitfield should be fixed-size would still be with us if we were to use a `bool` vec or a tightly-packed vec.

From what I can tell, a `bool` will almost always end up as a 8-bit value where 7 bits are wasted. Given that the minimum committee size is 128, if we store it as a `bool` we're definitely going to waste memory and I _believe_ we'd lose some time due to less efficient use of CPU cache. Like yourself, I'm not _hugely_ concerned about this but given we already have a tightly packed bit-vec impl and that abandoning it wouldn't solve the problems at hand I am tempted to stay with it. I might be missing something though?

**Comment by paulhauner** [comment]
Also, @ralexstokes you'll see there are some conflicts now. This would be due to #69 where `rustfmt` was run globally. You're safe to just override master with any of your changes.

They're simply formatting changes not logic changes. Sorry!

**Comment by rawfalafel** [comment]
Ah, I see. You’ve cleanly abstracted bit manipulation out. Makes sense!

**Comment by paulhauner** [comment]
Hey,
 I just went through both of those commits and they both look great!

In `18516c4` you're right about just setting the "one byte higher" bit to `false` instead of `true` then `false`.

I especially like the the `from_elem()` method. 

LGTM!

**Comment by ralexstokes** [comment]
@paulhauner this should be ready to go!

File: beacon_chain/validation/tests/attestation_validation/tests.rs [hunk 1]
```diff
@@ -129,16 +129,12 @@ fn test_attestation_validation_invalid_bad_bitfield_length() {
     /*
      * Extend the bitfield by one byte
      *
-     * This is a little hacky and makes assumptions about the internals
-     * of the bitfield.
+     * We take advantage of the fact that setting a bit outside the current bounds will grow the bitvector.
```
Comment (paulhauner) [inline beacon_chain/validation/tests/attestation_validation/tests.rs]: I love this re-wording! :rofl:

**Comment by paulhauner** [comment]
This looks great!

I didn't find any issues, but I produced a few more tests whilst I was reviewing. I have included them in a branch called [pr62-additional-tests](https://github.com/sigp/lighthouse/tree/pr62-additional-tests) for you to review/merge. They're not strictly necessary, just things that came to mind and I figure the more the merrier when it comes to unique tests. 

Notes on new tests:

- Added a test for a zero-length intialized bitfield. I was investigating `set()` (line 59) for integer wrapping. I didn't find any issues. I also renamed what you had as `test_empty_bitfield()` to `test_new_bitfield()` because I figure that name made more sense for the test I created.
- Added `test_grows_with_false()`. This ensures that when you set out-of-bounds further than 1 position you don't set all the intermediate newly created bits to `true`. I introduced this error when attempting to simplify the `set()` fn and noticed it didn't break any tests.  Turns out the `set()` fn didn't need simplification :)
- Added an extra scenario to `test_set_highest_bit()` where the highest set bit is the right-most bit.

I also derived `PartialEq` for `Error` because it was necessary for one of my new tests.

This is good to merge once you merge/reject those new tests :)

**Comment by ralexstokes** [comment]
these tests look great -- nice prodding :)

i'll merge them to this branch

**Comment by paulhauner** [comment]
Thanks, merging! This is going to make gitcoin bot soooo happy.

**Pull request merged**